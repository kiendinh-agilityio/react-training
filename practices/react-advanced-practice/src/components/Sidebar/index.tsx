import { memo, useMemo } from 'react';

// Import constants
import { SIDEBAR_ITEMS } from '@/constants';

// Import sidebar item
import SidebarItem from './SidebarItem';

// Import help box
import HelpBox from './HelpBox';

interface SidebarItemType {
  icon: React.ReactNode;
  label: string;
  active: boolean;
  href: string;
  group?: string;
}

const Sidebar = () => {
  const mainItems: SidebarItemType[] = useMemo(
    () => SIDEBAR_ITEMS.filter((item) => !item.group),
    [],
  );

  const accountItems: SidebarItemType[] = useMemo(
    () => SIDEBAR_ITEMS.filter((item) => item.group === 'ACCOUNT PAGES'),
    [],
  );

  // Function renderSidebarItems
  const renderSidebarItems = (items: SidebarItemType[]) =>
    items.map((item) => (
      <SidebarItem
        key={item.label}
        icon={item.icon}
        label={item.label}
        href={item.href}
        active={item.active}
      />
    ));

  return (
    <>
      <ul className="px-[30px]">
        {renderSidebarItems(mainItems)}
        <li className="font-bold font-helveticaBold py-6 px-4">ACCOUNT PAGES</li>
        {renderSidebarItems(accountItems)}
      </ul>
      <HelpBox />
    </>
  );
};

export default memo(Sidebar);
