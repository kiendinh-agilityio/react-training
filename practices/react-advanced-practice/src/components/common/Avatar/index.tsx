// import radix ui
import { Avatar as AvatarBase, AvatarProps } from '@radix-ui/themes';

type AvatarBaseProps = {
  ariaLabel?: string;
  alt?: string;
} & AvatarProps;

const Avatar = ({
  ariaLabel = 'Avatar',
  alt = 'Avatar',
  src = '',
  className = '',
  fallback,
}: AvatarBaseProps) => (
  <AvatarBase
    aria-label={ariaLabel}
    alt={alt}
    src={src}
    fallback={fallback}
    className={`rounded-base w-[40px] h-[40px] ${className}`}
  />
);

export default Avatar;
