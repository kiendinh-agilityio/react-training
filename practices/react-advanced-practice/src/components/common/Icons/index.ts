export { default as PurityIcon } from './PurityLogo';
export { default as SearchIcon } from './SearchIcon';
export { default as ShowPasswordIcon } from './ShowPassword';
export { default as HidePasswordIcon } from './HidePassword';
export { default as ArrowSelectIcon } from './ArrowSelect';
export { default as HomeIcon } from './HomeIcon';
export { default as ChartIcon } from './ChartIcon';
export { default as BillingIcon } from './BillingIcon';
export { default as SettingIcon } from './SettingIcon';
export { default as ProfileIcon } from './ProfileIcon';
export { default as SigninIcon } from './SigninIcon';
export { default as SignupIcon } from './SignupIcon';
export { default as DeleteIcon } from './DeleteIcon';
export { default as EditIcon } from './EditIcon';
export { default as DashboardIcon } from './DashboardIcon';
export { default as UserIcon } from './UserIcon';
export { default as LoginIcon } from './LoginIcon';
export { default as RegisterIcon } from './RegisterIcon';
export { default as AppleIcon } from './AppleIcon';
export { default as FacebookIcon } from './FacebookIcon';
export { default as GoogleIcon } from './GoogleIcon';
export { default as LoadingIcon } from './LoadingIcon';
export { default as PinionIcon } from './PinionIcon';
export { default as NotificationIcon } from './NotificationIcon';
export { default as QuestionIcon } from './QuestionIcon';
export { default as CloseIcon } from './CloseIcon';
export { default as FailedIcon } from './FailedIcon';
export { default as SuccessIcon } from './SuccessIcon';
export { default as TrashIcon } from './TrashIcon';
export { default as MoonIcon } from './MoonIcon';
export { default as SunIcon } from './SunIcon';
export { default as LogoutIcon } from './LogoutIcon';
