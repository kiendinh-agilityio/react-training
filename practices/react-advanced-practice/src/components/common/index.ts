export { default as Logo } from './Logo';
export { default as Heading } from './Heading';
export { default as Input } from './Input';
export { default as Select } from './Select';
