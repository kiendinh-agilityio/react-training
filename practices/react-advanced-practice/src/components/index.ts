export { default as Sidebar } from './Sidebar';
export { default as Profile } from './Profile';
export { default as AuthorTable } from './AuthorTable';
export { default as Navbar } from './Navbar';
export { default as SocialRegister } from './SocialRegister';
export { default as InputGroup } from './InputGroup';
export { default as AuthorForm } from './AuthorForm';
export { default as ConfirmModal } from './ConfirmModal';
export { default as SwitchTheme } from './SwitchTheme';
export { default as ErrorBoundary } from './ErrorBoundary';
export { default as AuthForm } from './AuthForm';
