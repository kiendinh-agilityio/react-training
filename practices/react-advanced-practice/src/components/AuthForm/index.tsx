import { useNavigate } from 'react-router-dom';
import { useState } from 'react';
import { useForm, SubmitHandler, Controller } from 'react-hook-form';
import { Flex } from '@radix-ui/themes';
import { Input, Button, Text, Link } from '@/components/common';
import {
  ShowPasswordIcon,
  HidePasswordIcon,
  LoadingIcon,
} from '@/components/common/Icons';
import { useAuthStore } from '@/stores';
import { REGEX, MESSAGE_ERROR } from '@/constants';

interface AuthFormInputs {
  name?: string;
  email: string;
  password: string;
}

interface AuthFormProps {
  type: 'signin' | 'signup';
  onSubmit: SubmitHandler<AuthFormInputs>;
  bottomText: string;
  bottomLink: string;
  bottomLinkText: string;
}

const AuthForm = ({
  type,
  onSubmit,
  bottomText,
  bottomLink,
  bottomLinkText,
}: AuthFormProps) => {
  const { showPassword, togglePasswordVisibility, setName, setEmail, setPassword } =
    useAuthStore();
  const navigate = useNavigate();
  const [isLoading, setIsLoading] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const {
    control,
    trigger,
    handleSubmit,
    formState: { errors },
  } = useForm<AuthFormInputs>({
    defaultValues: {
      name: '',
      email: '',
      password: '',
    },
  });

  /**
   * Creates an event handler for the blur event to validate a specific field.
   * @param fieldName - The name of the field to validate.
   * @param trigger - The react-hook-form trigger function.
   * @returns A function that triggers validation for the specified field.
   */
  const handleFieldBlur =
    (
      fieldName: keyof AuthFormInputs,
      trigger: (fieldName: keyof AuthFormInputs) => void,
    ) =>
    () =>
      trigger(fieldName);

  /**
   * Handles the onChange event for form inputs and triggers validation.
   * @param fieldName - The name of the field to validate.
   * @param fieldOnChange - The react-hook-form onChange handler for the field.
   * @returns A function that handles the input's onChange event.
   */
  const handleFieldChange =
    (fieldName: keyof AuthFormInputs, fieldOnChange: (value: string) => void) =>
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      fieldOnChange(value);
      trigger(fieldName);

      fieldName === 'password' && value.match(REGEX.PASSWORD) && setErrorMessage('');
    };

  // Handle Form Submit
  const handleFormSubmit: SubmitHandler<AuthFormInputs> = async (data) => {
    setIsLoading(true);
    setErrorMessage('');

    try {
      if (type === 'signup') {
        setName(data.name || '');
        setEmail(data.email);
        setPassword(data.password);
      }

      await onSubmit(data);

      setTimeout(() => {
        navigate('/home');
      }, 1500);
    } catch (error) {
      setIsLoading(false);
      setErrorMessage(
        type === 'signin' ? MESSAGE_ERROR.INVALID_SIGNIN : MESSAGE_ERROR.SIGNUP_FAILED,
      );
    }
  };

  return (
    <form
      onSubmit={handleSubmit(handleFormSubmit)}
      className="flex flex-col max-w-[350px]"
    >
      {type === 'signup' && (
        <Flex direction="column" className="mb-6 gap-[5px]">
          <Text>Name</Text>
          <Controller
            name="name"
            control={control}
            rules={{
              required: MESSAGE_ERROR.REQUIRED_ERROR('Name'),
              pattern: {
                value: REGEX.NAME,
                message: MESSAGE_ERROR.INVALID_NAME,
              },
            }}
            render={({ field }) => (
              <Input
                {...field}
                placeholder="Your full name"
                type="text"
                errorMessage={errors.name?.message}
                onChange={handleFieldChange('name', field.onChange)}
                onBlur={handleFieldBlur('name', trigger)}
              />
            )}
          />
        </Flex>
      )}

      <Flex direction="column" className="mb-6 gap-[5px]">
        <Text>Email</Text>
        <Controller
          name="email"
          control={control}
          rules={{
            required: MESSAGE_ERROR.REQUIRED_ERROR('Email'),
            pattern: {
              value: REGEX.EMAIL,
              message: MESSAGE_ERROR.INVALID_EMAIL,
            },
          }}
          render={({ field }) => (
            <Input
              {...field}
              placeholder="Your email address"
              type="email"
              errorMessage={errors.email?.message}
              onChange={handleFieldChange('email', field.onChange)}
              onBlur={handleFieldBlur('email', trigger)}
            />
          )}
        />
      </Flex>

      <Flex direction="column" className="gap-[5px]">
        <Text>Password</Text>
        <Controller
          name="password"
          control={control}
          rules={{
            required: MESSAGE_ERROR.REQUIRED_ERROR('Password'),
            pattern: {
              value: REGEX.PASSWORD,
              message: MESSAGE_ERROR.INVALID_PASSWORD,
            },
          }}
          render={({ field }) => (
            <Input
              {...field}
              placeholder="Your password"
              type={showPassword ? 'text' : 'password'}
              errorMessage={errors.password?.message}
              onChange={handleFieldChange('password', field.onChange)}
              onBlur={handleFieldBlur('password', trigger)}
              rightIcon={
                <button
                  type="button"
                  onClick={togglePasswordVisibility}
                  aria-label={showPassword ? 'Hide password' : 'Show password'}
                >
                  {showPassword ? <ShowPasswordIcon /> : <HidePasswordIcon />}
                </button>
              }
            />
          )}
        />
      </Flex>

      {errorMessage && !errors.password && (
        <Text className="text-danger mt-3 text-sm">{errorMessage}</Text>
      )}

      <Button
        onClick={handleSubmit(handleFormSubmit)}
        className="mt-9 w-[350px]"
        isDisabled={isLoading}
      >
        {isLoading ? <LoadingIcon /> : type === 'signin' ? 'Sign In' : 'Sign Up'}
      </Button>
      <Text as="div" className="font-regular text-base text-center mt-[22px]">
        {bottomText}{' '}
        <Link href={bottomLink} className="text-primary">
          {bottomLinkText}
        </Link>
      </Text>
    </form>
  );
};

export default AuthForm;
