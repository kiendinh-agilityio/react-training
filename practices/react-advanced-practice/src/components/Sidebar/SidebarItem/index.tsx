import { memo } from 'react';

// Import components
import { IconWrapper } from '@/components/common/';

interface SidebarItemProps {
  icon: React.ReactNode;
  label: string;
  active: boolean;
  href: string;
}

const SidebarItem = memo(({ icon, label, active, href }: SidebarItemProps) => {
  const className = `sidebar-item ${active ? 'bg-white text-dark w-[220px] h-[54px] flex items-center rounded-[15px] cursor-pointer' : ''}`;

  return (
    <li className={className}>
      <a className="flex items-center gap-4" href={href}>
        <IconWrapper icon={icon} active={active} />
        <span className="text-xs leading-base">{label}</span>
      </a>
    </li>
  );
});

export default SidebarItem;
