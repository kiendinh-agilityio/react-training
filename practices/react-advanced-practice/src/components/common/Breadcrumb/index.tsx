// Import radix ui
import { Box } from '@radix-ui/themes';

// Import common components
import { Text } from '@/components/common';

interface BreadcrumbProps {
  currentPage: string;
  label: string;
}

const Breadcrumb = ({ currentPage, label }: BreadcrumbProps) => (
  <Box>
    <Text className="flex gap-[4px] text-base">
      {label} /{' '}
      <Box as="span" className="text-xs text-dark dark:text-light">
        {currentPage}
      </Box>
    </Text>
  </Box>
);

export default Breadcrumb;
