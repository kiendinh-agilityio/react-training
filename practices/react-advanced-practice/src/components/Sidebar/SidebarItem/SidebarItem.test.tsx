import { render, screen, fireEvent } from '@testing-library/react';
import SidebarItem from '.';
import { IconWrapper } from '@/components/common/';
import { DashboardIcon } from '@/components/common/Icons';

// Mock IconWrapper component
jest.mock('@/components/common/', () => ({
  IconWrapper: jest.fn(({ icon }) => <div data-testid="icon-wrapper">{icon}</div>),
}));

describe('SidebarItem', () => {
  const defaultProps = {
    icon: <DashboardIcon />,
    label: 'Home',
    active: false,
    href: '/home',
  };

  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(<SidebarItem {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with correct label and link', () => {
    render(<SidebarItem {...defaultProps} />);
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByRole('link')).toHaveAttribute('href', '/home');
  });

  it('renders active class when active is true', () => {
    render(<SidebarItem {...defaultProps} active={true} />);
    expect(screen.getByRole('listitem')).toHaveClass('bg-white text-dark');
  });

  it('applies hover styles on mouse enter', () => {
    render(<SidebarItem {...defaultProps} />);
    const listItem = screen.getByRole('listitem');

    fireEvent.mouseEnter(listItem);
    expect(listItem).toHaveClass('bg-white text-dark');

    fireEvent.mouseLeave(listItem);
    expect(listItem).not.toHaveClass('bg-white text-dark');
  });

  it('passes correct props to IconWrapper', () => {
    render(<SidebarItem {...defaultProps} active={true} />);
    expect(IconWrapper).toHaveBeenCalledWith(
      expect.objectContaining({ active: true }),
      expect.any(Object),
    );
  });
});
