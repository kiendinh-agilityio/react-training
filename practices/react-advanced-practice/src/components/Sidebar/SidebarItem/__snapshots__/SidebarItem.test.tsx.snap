// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarItem renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="py-2 px-4 text-sm w-[220px] h-[54px] rounded-[15px] font-bold text-base mb-3 "
  >
    <a
      class="flex items-center gap-4"
      href="/home"
    >
      <div
        data-testid="icon-wrapper"
      >
        <svg
          fill="#4fd1c5"
          height="12"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#a)"
            fill="#2d3748"
          >
            <path
              d="M9.972 3.675a.086.086 0 0 0 0-.148L6.691 1.62a1.378 1.378 0 0 0-1.382 0L2.028 3.527a.086.086 0 0 0 0 .148l3.929 2.312a.086.086 0 0 0 .087 0l3.928-2.312ZM1.66 4.26a.086.086 0 0 0-.129.075V8.07a1.031 1.031 0 0 0 .512.889l3.484 2.089a.086.086 0 0 0 .13-.074V6.64a.086.086 0 0 0-.044-.074L1.66 4.26ZM6.344 6.655v4.318a.086.086 0 0 0 .129.075l3.483-2.09a1.031 1.031 0 0 0 .513-.887V4.335a.085.085 0 0 0-.043-.074.086.086 0 0 0-.086 0L6.387 6.58a.086.086 0 0 0-.043.074Z"
            />
          </g>
          <defs>
            <clippath
              id="a"
            >
              <path
                d="M.5.748h11v11H.5z"
                fill="#fff"
              />
            </clippath>
          </defs>
        </svg>
      </div>
      <span
        class="text-xs leading-base"
      >
        Home
      </span>
    </a>
  </li>
</DocumentFragment>
`;
